cmd/logger.go

```diff
@@ -17,141 +17,219 @@
-}{}
-
-// logger carries logging configuration for various supported loggers.
-// Currently supported loggers are
-//
-//   - console [default]
-//   - file
-type logger struct {
+type loggers struct {
 	sync.RWMutex
-	Console consoleLogger `json:"console"`
-	File    fileLogger    `json:"file"`
-	// Add new loggers here.
+	Console ConsoleLogger `json:"console"`
+	File    FileLogger    `json:"file"`
 }
 
-/// Logger related.
-
-// Validate logger contents
-func (l *logger) Validate() error {
-	if l == nil {
-		return nil
-	}
-	if err := l.Console.Validate(); err != nil {
-		return fmt.Errorf("`Console` field: %s", err.Error())
+// Validate - Check whether loggers are valid or not.
+func (l *loggers) Validate() (err error) {
+	if l != nil {
+		fileLogger := l.GetFile()
+		if fileLogger.Enable && fileLogger.Filename == "" {
+			err = errors.New("Missing filename for enabled file logger")
+		}
 	}
-	if err := l.File.Validate(); err != nil {
-		return fmt.Errorf("`File` field: %s", err.Error())
-	}
-	return nil
+
+	return err
 }
 
 // SetFile set new file logger.
-func (l *logger) SetFile(flogger fileLogger) {
+func (l *loggers) SetFile(flogger FileLogger) {
 	l.Lock()
 	defer l.Unlock()
 	l.File = flogger
 }
 
 // GetFileLogger get current file logger.
-func (l *logger) GetFile() fileLogger {
+func (l *loggers) GetFile() FileLogger {
 	l.RLock()
 	defer l.RUnlock()
 	return l.File
 }
 
 // SetConsole set new console logger.
-func (l *logger) SetConsole(clogger consoleLogger) {
+func (l *loggers) SetConsole(clogger ConsoleLogger) {
 	l.Lock()
 	defer l.Unlock()
 	l.Console = clogger
 }
 
-func (l *logger) GetConsole() consoleLogger {
+// GetConsole get current console logger.
+func (l *loggers) GetConsole() ConsoleLogger {
 	l.RLock()
 	defer l.RUnlock()
 	return l.Console
 }
 
-// Get file, line, function name of the caller.
-func callerSource() string {
-	pc, file, line, success := runtime.Caller(2)
-	if !success {
-		file = "<unknown>"
-		line = 0
+// LogTarget - interface for log target.
+type LogTarget interface {
+	Fire(entry *logrus.Entry) error
+	String() string
+}
+
+// BaseLogTarget - base log target.
+type BaseLogTarget struct {
+	Enable    bool     `json:"enable"`
+	Level     LogLevel `json:"level"`
+	formatter logrus.Formatter
+}
+
+// Logger - higher level logger.
+type Logger struct {
+	logger        *logrus.Logger
+	consoleTarget ConsoleLogger
+	targets       []LogTarget
+	quiet         bool
+}
+
+// AddTarget - add logger to this hook.
+func (log *Logger) AddTarget(logTarget LogTarget) {
+	log.targets = append(log.targets, logTarget)
+}
+
+// SetConsoleTarget - sets console target to this hook.
+func (log *Logger) SetConsoleTarget(consoleTarget ConsoleLogger) {
+	log.consoleTarget = consoleTarget
+}
+
+// Fire - log entry handler to save logs.
+func (log *Logger) Fire(entry *logrus.Entry) (err error) {
+	if err = log.consoleTarget.Fire(entry); err != nil {
+		log.Printf("Unable to log to console target. %s\n", err)
```

Comment:
Shouldn't we return here, since we found an error?

Otherwise, the next for block over-writes the error we found here.